File: src/main/scala/gameover/fwk/pool/ObjectPool.scala
package gameover.fwk.pool

import com.badlogic.gdx.utils.Pool
import gameover.fwk.libgdx.utils.LibGDXHelper

import scala.reflect.ClassTag

trait ObjectPool[T] extends LibGDXHelper {

  private val pool: Pool[T] = new Pool[T]() {
    protected def newObject: T = instantiateObject
  }

  def instantiateObject: T

  def clear() {
    pool.clear()
  }

  def free(a: Array[T]) {
    a.foreach(free)
  }

  def free(a: GdxArray[T]) {
    a.foreach(free)
  }

  def free(l: List[T]) {
    l.foreach(free)
  }

  def free[U: ClassTag](t: T) {
    pool.free(t)
  }

  def obtain(): T = pool.obtain()

  def obtainAsArray[T:ClassTag](nb: Int): Array[T] = Array.fill[T](nb)(obtain().asInstanceOf[T])

  def obtainAsGdxArray(nb: Int): GdxArray[T] =  new GdxArray[T]().fill(nb)(obtain)

  def obtainAsList(nb: Int): List[T] = List.fill(nb)(obtain())
}

File: src/main/kotlin/gameover/fwk/pool/ObjectPool.kt
package gameover.fwk.pool

import com.badlogic.gdx.utils.Pool
import com.badlogic.gdx.math.GridPoint2
import com.badlogic.gdx.math.Rectangle
import com.badlogic.gdx.math.Vector2
import gameover.fwk.libgdx.utils.GdxArray

import java.util.*

abstract class ObjectPool<T> {

    abstract fun instantiateObject(): T

    private val pool = object: Pool<T>() {
        override fun newObject(): T { return instantiateObject() }
    }

    /**
     * Clear all objects in pool.
     */
    fun clear() {
        pool.clear()
    }

    fun free(a: Array<T>) {
        for (e in 0..(a.size - 1))
            free(a.get(e))
    }

    fun free(a: GdxArray<T>) {
        for (e in 0..(a.size - 1))
            free(a.get(e))
    }

    fun free(l: List<T>) {
        for (e in l)
            free(e)
    }

    fun free(t: T) {
        pool.free(t)
    }

    fun obtain(): T = pool.obtain()

    abstract fun obtainAsArray(nb: Int) : Array<T>

    fun obtainAsGdxArray(nb: Int): GdxArray<T> {
        val a = GdxArray<T>(nb)
        for (i in 1..nb)
            a.add(obtain())
        return a
    }

    fun obtainAsList(nb: Int): ArrayList<T> {
        val a = ArrayList<T>(nb)
        for (i in 1..nb)
            a.add(obtain())
        return a
    }
}

object GridPoint2Pool : ObjectPool<GridPoint2>() {

    override fun instantiateObject() : GridPoint2 = GridPoint2()

    fun obtain(o: GridPoint2): GridPoint2 = obtain().set(o)

    fun obtain(x: Int, y: Int): GridPoint2 = obtain().set(x, y)

    override fun obtainAsArray(nb: Int) : Array<GridPoint2> = Array(nb) {
        obtain()
    }
}

object RectanglePool : ObjectPool<Rectangle>() {

    override fun instantiateObject(): Rectangle = Rectangle()

    fun obtain(rectangle: Rectangle): Rectangle = obtain().set(rectangle)

    fun obtain(v1: Vector2, v2: Vector2): Rectangle = obtain().set(v1.x, v1.y, v2.x, v2.y)

    fun obtain(x1: Float, y1: Float, w: Float, h: Float): Rectangle = obtain().set(x1, y1, w, h)

    override fun obtainAsArray(nb: Int) : Array<Rectangle> = Array(nb) {
        obtain()
    }
}

object Vector2Pool : ObjectPool<Vector2>() {

    override fun instantiateObject(): Vector2 = Vector2()

    fun obtain(o: Vector2): Vector2 = obtain().set(o)

    fun obtain(x: Float, y: Float): Vector2 = obtain().set(x, y)

    override fun obtainAsArray(nb: Int) : Array<Vector2> = Array(nb) {
        obtain()
    }
}